### ISSUE TITLE
neo4j-driver v1 10 times slower response times compared with python driver or Graphene DB web client

### ISSUE THREAD

**Issue by cameliaben** [neo4j/neo4j-javascript-driver#440]
Hello,

I am using the **javascript neo4j-driver** ( v1 )for an API written with Node.js to connect to a database hosted on GrapheneDb (shared server plan), Neo4j version listed on GrapheneDb client is 3.3.7, and the database consists of around:
- **Nodes**: 158193
- **Labels**: 84
- **Relationships**: 3735575
- **Relationship types**: 137

**The problem that I struggle with is very high latency when I query the database from the API using neo4j-driver, compared to querying the same database using  GrapheneDb client OR a python test script using python driver.** 
**_Queries that take under  0.3 seconds on GrapheneDB client or using pyhton driver, take more than 3 seconds sometimes using the neo4j-driver._** It never occurs with neo4j-driver to have identical response times compared with python driver at least. This problem gets bigger and bigger for complex queries, that sometimes even on GrapheneDB take 2 -3 seconds and thus when executed from the API they take forever.

**Details about the python driver used to compare response times:** 
- Anaconda-based environment with Python 3.6.8
- neo4j module 1.7.1 (installed via pip)

**Details about the neo4j-driver setup and usage:**
`const neo4j = require( "neo4j-driver" ).v1;`

`const config = {
    logger: neo4j.logging.console( "debug" ),
    maxConnectionLifetime: 20 * 60 * 1000, 
    connectionTimeout: 1000 * 45,
    connectionAcquisitionTimeout: 60 * 1000, 
    maxTransactionRetryTime: 10 * 1000,
    maxConnectionPoolSize: 300,
};`

`const myDriver = neo4j.driver(
    "bolt://url",
    neo4j.auth.basic( "db_user", "db_pass" ),
    config );`

There is only one driver instance per application, and then  a promise-based approach is used to execute queries (there are many cases when  such queries run in parallel. Sequential execution led significantly to worse response times.) :

`const getResultsFromDb = ( myParameter) => {
    const myParameterizedQuery= "my_query_text_using_paramater";`

    const param = { myParameter };
    const session = driver.session();

    return session.writeTransaction( tx => tx.run( myParameterizedQuery, param ) )
        .then( ( result ) => {
            session.close();
            const processedResult = extractValuesAtStartingIndex( result, 0 );
            return processedResult ;
        } )
        .catch( ( error ) => {
            session.close();
            return { error };
        } );
};`

I am starting to think if I would benefit more from taking the streaming-approach, as documented here: 
https://www.npmjs.com/package/neo4j-driver#usage-examples
but this involves a lot of refactoring at this point, since the entire API is written using promises and I am not sure that it would fit my apps needs. 
 I couldn't find in the neo4j-driver documentation pros and cons (in terms of performance) for any of the 2 approaches.

**What I have tried so far, to improve response times:**
- fine tune my queries, which is still work in progress. But no huge leaps here that would take my response time below 1 second, this is my final goal.
- improve the data model by adding new labels to limit db hits => some performance gains here, sometimes even 50% improvements from queries taking ~20 seconds to taking < ~10 seconds

**I would appreciate  a lot if someone would look into this issue and help me understand:**
1. what am I doing wrong / am I missing something with my approach ?
2. **is there something specific to neo4j-driver JS that causes such high response times compared to the other methods (pyhton driver, GrapheneDB web client) ?**
3. should I totally move away from the promise based approach, and take the streaming-based one instead ?
4.  what would be the best approach for an app running in production and using neo4j-driver ?

**Thank you in advance.**

**Comment by zhenlineo** [comment]
Hi @cameliaben,

Can you print out `result summary` after iterating all records? We can check the time spent on query execution and on streaming data.

```
return session.writeTransaction( tx => tx.run( myParameterizedQuery, param ) )
    .then( ( result ) => {
        const processedResult = extractValuesAtStartingIndex( result, 0 );
        console.log(result.summary()); // such as adding the printout here.
        session.close(); // this call enforces to pull all unconsumed records into memory.
        return processedResult ;
    } )
    .catch( ( error ) => {
        session.close();
        return { error };
    } );
```

Then try to print the result summary with python driver too.
(You might want to repeat several times until the number becomes stable as Neo4j caches previous db lookups result in memory.)

**Comment by cameliaben** [comment]
Hi @zhenlineo ,

**For a query that takes around  1245 ms when executed using the javascript neo4j-driver the result summary looks like this :**

`{
  "statement": {
    "text": "\"my_query\"",
    "parameters": {}
  },
  "statementType": "r",
  "counters": {
    "_stats": {
      "nodesCreated": 0,
      "nodesDeleted": 0,
      "relationshipsCreated": 0,
      "relationshipsDeleted": 0,
      "propertiesSet": 0,
      "labelsAdded": 0,
      "labelsRemoved": 0,
      "indexesAdded": 0,
      "indexesRemoved": 0,
      "constraintsAdded": 0,
      "constraintsRemoved": 0
    }
  },
  "updateStatistics": {
    "_stats": {
      "nodesCreated": 0,
      "nodesDeleted": 0,
      "relationshipsCreated": 0,
      "relationshipsDeleted": 0,
      "propertiesSet": 0,
      "labelsAdded": 0,
      "labelsRemoved": 0,
      "indexesAdded": 0,
      "indexesRemoved": 0,
      "constraintsAdded": 0,
      "constraintsRemoved": 0
    }
  },
  "plan": false,
  "profile": false,
  "notifications": [],
  "server": {
    "address": "\"my_db_url\"",
    "version": "Neo4j/3.3.7"
  },
  "resultConsumedAfter": {
    "low": 1,
    "high": 0
  },
  "resultAvailableAfter": {
    "low": 387,
    "high": 0
  }
}`

I measured the execution time for JS neo4j-driver using two variables, a `var start = Date.now();` that I declare right before query execution, and on .then( ), when the result comes back, I basically get the time difference: `const duration = Date.now() - start; `

**For the same query, python driver has the following result summary and takes under 1 second:**
`{'counters': {},
 'metadata': {'fields': ['a'],
              'parameters': {},
              'protocol_version': 2,
              'result_available_after': 277,
              'result_consumed_after': 1,
              'server': '<neobolt.direct.ServerInfo object at 0x000001F21F4C7DD8>',
              'statement': 'query_text',
              'type': 'r'},
 'notifications': [],
 'parameters': {},
 'protocol_version': 2,
 'result_available_after': 277,
 'result_consumed_after': 1,
 'server': '<neobolt.direct.ServerInfo object at 0x000001F21F4C7DD8>',
 'statement': 'query_text',
 'statement_type': 'r',
 't_first': None,
 't_last': None}`

Thank you for looking into this.

**Comment by zhenlineo** [comment]
Hi @cameliaben,

In the printout, `resultAvailableAfter` describes the time spent on the database to run the query. `resultConsumedAfter` states the time spent on server to send the whole result stream. These numbers are a server side only measurement, as a result, it should not vary for the same query with different drivers.

We can see that in the run with JS driver, the time spent on `resultAvailableAfter` is slightly longer than Python driver. If this is not related to drivers, then why the database spent longer time to run with JS driver but shorter with python driver? It could be that the database is not properly warmed up. It could be that the JS driver consumes too much computing resources if the database and the driver are on the same machine. Or some other reasons. 

Then for the other 700ms spent both for JS driver and python driver, could it be network delay if you are hosting Neo4j in a cloud service?

It is hard to compare performance among different language drivers. Python driver pre-complied some of its code to C for Linux. If you are running test on a Linux machine, you might actually comparing JS script with pre-compiled C code.

As the time difference is not as big as 10x, so I do not think there is a potential bug in JS driver.

Here are some suggestions how you can improve the performance for your app with Js driver:
1) Adding labels to reduce the query execution time. No matter what driver you are using, we cannot be faster than this query execution time.
2) Do not write `session.close()` or `transaction.close()` before you actually have handled all records in the query result. As `session.close()` or `transaction.close()` will block until the whole result has fully arrived on the client (and pull the result into memory). 
3) Using `subscribe` (streaming approach) if possible. Though I do not think it will make a big difference than the other API if you've followed suggestion 2). But it helps to write correct code. 

Cheers,
Zhen

**Comment by cameliaben** [comment]
@zhenlineo ,

Thank you for your suggestions. 
I will definitely try suggestion number 2), and suggestion 1) is something that is a work in progress, under constant improvement.

As for number 3), at least at this point where the API is mature and it would require a lot of refactoring, it will be postponed but kept in mind for future changes.

Thank you once again for looking into this.

**Comment by cameliaben** [comment]
@zhenlineo  - Hello again,

I think only just now I really understood your second comment above, the part related to `resultConsumedAfter` .

I have started to take a stream-based approach as to how database results are consumed by my Node.js server. With this approach, from my initial tests, it looks like I am getting closer and closer with total execution time to the value specified in `resultConsumedAfter` variable, that is coming back from the database server. 
Without having updated my problematic queries and without changing the database structure , it seems that this time value is the maximum that I can get in terms of total response time for an executed query, with Node.js.

From your knowledge, do you know any way to lower the `resultConsumedAfter`, aside from query tuning / database model updates ?

Thank you!

**Comment by zhenlineo** [comment]
Hi @cameliaben,

No. It is the cypher execution time. If you really really need a faster reply, then you can try working on kernel API directly using [Procedures](https://neo4j.com/docs/java-reference/current/extending-neo4j/procedures/). It might save you some query parsing and planning time.

**Comment by technige** [comment]
Closing issue as nothing to fix.
